### PR TITLE
spec: Introduce `/farmer` endpoints, add authentication, improve `GET /login` 

### PR DESCRIPTION
**Pull request by xdustinface**
Specification of the changes added by #75

### CODE REVIEW

File: SPECIFICATION.md [hunk 1]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
 
 Successful response:
 ```json
-{"points_balance": 1130, "current_difficulty": 10}
+{"welcome_message" : "Welcome to the reference pool. Happy farming."}
 ```
 
-The successful response must always contain the points balance since the last payout, as well as the current difficulty
-that this farmer is being given credit for.
+The successful response must always contain a welcome message which must be defined by the pool.
```
Comment (aqk) [inline SPECIFICATION.md:212]: Do we care about max size of the welcome message? @mariano54

File: SPECIFICATION.md [hunk 2]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+#### payload.launcher_id
+The unique identifier of the farmer's singleton, see [Farmer identification](#farmer-identification).
+
+#### payload.authentication_token
+See [Farmer authentication](#farmer-authentication) for the specification of
+`authentication_token`.
+
+#### payload.authentication_public_key
+The public key of the authentication key, which is a temporary key used by the farmer to sign requests
+to the pool. It is an authorization given by the `owner_key`, so that the owner key can be potentially kept more secure.
```
Comment (aqk) [inline SPECIFICATION.md:225]: `It is an authorization given by` -> `It is authorized by`

Comment (aqk) [inline SPECIFICATION.md:225]: `can be potentially kept more secure` -> `can be kept more secure`

("more" is enough here)

File: SPECIFICATION.md [hunk 3]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+The pool should reject requests made with outdated `authentication_keys`. This key can be changed using `PUT /farmer`,
```
Comment (aqk) [inline SPECIFICATION.md:226]: `This keys can` -> `These keys can`

File: SPECIFICATION.md [hunk 4]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+which is signed with the owner key.
 
-Failed response:
+#### payload.payout_instructions
+This is the instructions for how the farmer wants to get paid. By default this will be an XCH address, but it can
```
Comment (aqk) [inline SPECIFICATION.md:230]: `This is the instructions` -> `These are the instructions`

File: SPECIFICATION.md [hunk 5]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+be set to any string with a size of less than 1024 characters, so it can represent another blockchain or payment
+system identifier.
+
+#### payload.suggested_difficulty
+A request from the farmer to update the difficulty. Can be ignored or respected by the pool. However, this should only
```
Comment (aqk) [inline SPECIFICATION.md:235]: We should say here what it means for this to be lowered or increased.

File: SPECIFICATION.md [hunk 6]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+        "authentication_public_key": "0x970e181ae45435ae696508a78012dc80548c334cf29676ea6ade7049eb9d2b9579cc30cb44c3fd68d35a250cfbc69e29",
+        "payout_instructions": "0xc2b08e41d766da4116e388357ed957d04ad754623a915f3fd65188a8746cf3e8",
+        "suggested_difficulty": 10
+    },
+    "signature": "0xa078dc1462bbcdec7cd651c5c3d7584ac6c6a142e049c7790f3b0ee8768ed6326e3a639f949b2293469be561adfa1c57130f64334994f53c1bd12e59579e27127fbabadc5e8793a2ef194a5a22ac832e92dcb6ad9a0d33bd264726f6e8df6aad"
+}
+```
+
+For a description of the request body entries see the corresponding keys in [POST /farmer](#post-farmer). The values
+provided with the key/value pairs are used to update the existing values. All entries, except `launcher_id`, are
```
Comment (aqk) [inline SPECIFICATION.md:269]: `the existing values` -> `the existing values on the server`

File: SPECIFICATION.md [hunk 7]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+Successful response:
+```json
+{
+  "authentication_public_key": true,
+  "payout_instructions": true,
+  "suggested_difficulty": true
+}
+```
+
+The successful response must always contain one key/value pair for each entry provided in the request body. The value
```
Comment (aqk) [inline SPECIFICATION.md:286]: `The successful response` -> `A successful response`

I think there is one case that English and German differ!

File: SPECIFICATION.md [hunk 8]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
+}
+```
+
+Successful response:
+```json
+{ "new_difficulty": 10}
 ```
-Failed responses must include an error code as well as an english error message.
+
+The successful response must always contain the new difficulty which should be respected by the farmer.
```
Comment (aqk) [inline SPECIFICATION.md:331]: The successful -> A successful

Comment (aqk) [inline SPECIFICATION.md:331]: should be respected -> must be  respected

File: SPECIFICATION.md [hunk 9]
```diff
@@ -63,61 +123,223 @@ The minimum difficulty that the pool supports. This will also be the default tha
...
 
 #### payload
 This is the main payload of the partial, which is signed by two keys: `authentication_key` and `plot_key`.
```
Comment (aqk) [inline SPECIFICATION.md:334]: Might want to say how - is this a BLS aggsig?

File: SPECIFICATION.md [hunk 10]
```diff
@@ -126,7 +348,7 @@ The challenge of the proof of space, computed from the signage point or end of s
 
 #### payload.proof_of_space.pool_contract_puzzle_hash
 The puzzle hash that is encoded in the plots, equivalent to the `p2_singleton_puzzle_hash`. This is the first place
```
Comment (aqk) [inline SPECIFICATION.md:350]: first place -> first address

File: SPECIFICATION.md [hunk 11]
```diff
@@ -139,92 +361,82 @@ K size, must be at least 32.
...
-(This section is a draft and open to feedback)
+This allows the user to log in to a web interface if the pool supports it, see service flags in
+[GET /pool_info](#get-pool_info). The format of the request is the following:
+
+The farmer software must offer a way to generate and display a login link or provide a button which generates the
+link and then just opens it in the default browser. The link follow the specification below.
 
-This allows the user to log in to a web interface if the pool supports it. This is optional in V1.
-The format of the request is the following:
+Note that there is no explicit account creation. A farmer can log in after making their self known at the pool with
```
Comment (aqk) [inline SPECIFICATION.md:397]: Great Note!

File: SPECIFICATION.md [hunk 12]
```diff
@@ -139,92 +361,82 @@ K size, must be at least 32.
...
-This is a code generated by the pool, and shown to the user. It can be any alphanumeric string.
-This must have a length <= 32 bytes.
+#### authentication_token
+See [Farmer authentication](#farmer-authentication) for the specification of
+`authentication_token`.
 
-#### authentication_pk
-The hex representation of the authentication public key described above.
+#### target_puzzle_hash
+The pools target puzzle hash, see [GET /pool_info](#get-pool_info)
```
Comment (aqk) [inline SPECIFICATION.md:421]: pools -> pool's 
target puzzle hash -> payout address

**Review by mariano54 — DISMISSED** [review]

File: SPECIFICATION.md [hunk 13]
```diff
@@ -1,57 +1,117 @@
...
-The parties involved in the pool protocol involve the pool operator and farmers. Each farmer is running
+The parties involved in the pool protocol are the pool operator and farmers. Each farmer is running
 a farmer process, and any number of harvester processes connected to that farmer process. The full node can either be
 run by the farmer (the default in the Chia GUI application), or run by the pool operator. If the farmer does not want
 to run a full node, they can configure their node to connect to a remote full node.
 
-A pool operator can support any number of farmers. 
+A pool operator can support any number of farmers.
+
+## Farmer identification
```
Comment (aqk) [inline SPECIFICATION.md:17]: We should add a few sentences to make explicit the difference between this part of the pool protocol, and the optional `/login` below.

I think `/login` is described well enough once they get down that far in the document.

**Review by aqk — COMMENTED** [review]
Looks really good. I have a few comments, mostly small grammar changes.

**Comment by xdustinface** [comment]
@aqk Thanks! See c22ef0b, i hope i addressed everything.

**Pull request merged**